gateway/ln-gateway/src/gateway_module_v2/send_sm.rs

```diff
@@ -0,0 +1,202 @@
+pub struct SendStateMachine {
+    pub operation_id: OperationId,
+    pub contract: OutgoingContract,
+    pub state: SendSMState,
+}
+
+#[allow(clippy::large_enum_variant)]
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Decodable, Encodable)]
+pub enum SendSMState {
+    Sending(Sending),
+    Claiming(Claiming),
+    Cancelled,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Decodable, Encodable)]
+pub struct Sending {
+    pub max_delay: u64,
+    pub max_fee_msat: u64,
+    pub invoice: Bolt11Invoice,
+    pub claim_keypair: KeyPair,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Decodable, Encodable)]
+pub struct Claiming {
+    pub(crate) preimage: [u8; 32],
+    outpoints: Vec<OutPoint>,
+}
+
+#[cfg_attr(doc, aquamarine::aquamarine)]
+/// State machine that handles the relay of an incoming Lightning payment.
+///
+/// ```mermaid
+/// graph LR
+/// classDef virtual fill:#fff,stroke-dasharray: 5 5
+///
+///     Sending -- payment is successful --> Claiming
+///     Sending -- payment fails --> Cancelled
+/// ```
+impl State for SendStateMachine {
+    type ModuleContext = GatewayClientContextV2;
+
+    fn transitions(
+        &self,
+        context: &Self::ModuleContext,
+        global_context: &DynGlobalClientContext,
+    ) -> Vec<StateTransition<Self>> {
+        let gc = global_context.clone();
+
+        match &self.state {
+            SendSMState::Sending(sending) => {
+                vec![StateTransition::new(
+                    Self::send_payment(
+                        context.clone(),
+                        sending.max_delay,
+                        sending.max_fee_msat,
+                        sending.invoice.clone(),
+                    ),
+                    move |dbtx, result, old_state| {
+                        Box::pin(Self::transition_send_payment(
+                            dbtx,
+                            old_state,
+                            gc.clone(),
+                            result,
+                        ))
+                    },
+                )]
+            }
+            SendSMState::Claiming(..) => {
+                vec![]
+            }
+            SendSMState::Cancelled => {
+                vec![]
+            }
+        }
+    }
+
+    fn operation_id(&self) -> OperationId {
+        self.operation_id
+    }
+}
+
+impl SendStateMachine {
+    async fn send_payment(
+        context: GatewayClientContextV2,
+        max_delay: u64,
+        max_fee_msat: u64,
+        invoice: Bolt11Invoice,
+    ) -> Result<[u8; 32], ()> {
+        let lightning_context = loop {
+            match context.gateway.get_lightning_context().await {
+                Ok(lightning_context) => break lightning_context,
+                Err(_) => {
+                    sleep(std::time::Duration::from_secs(1)).await;
+                }
+            }
+        };
+
+        if lightning_context.lightning_public_key == invoice.recover_payee_pub_key() {
```

Comment:
I think this check assumes that a lightning node and a gateway are 1 to 1. This might not be true in the future.

Reply:
Can we add a comment here?